### ISSUE TITLE
Froala Editor 3.x Support

### ISSUE THREAD

**Issue by Panman82** [froala/ember-froala-editor#73]
Now that version 3 is out, this addon should pull in that version.

There are larger, major version changes that I'd also like to work through before releasing a v3 here. This includes Ember Octane stuff such as moving to Glimmer Components. Look at the [3.0 milestone](https://github.com/froala/ember-froala-editor/milestone/1) for those.

**Comment by Panman8201** [comment]
Update; I had been working on this through the beginning of June but go pulled off due to other work duties (and summer family fun time). Hoping to get back to this soon, since Ember 3.11 includes many of the pieces I need to move this to Octane.

**Comment by Panman8201** [comment]
Update; Sorry for the delay here, still very excited to get this addon to Octane Edition. Unfortunately work and family time have gotten in the way. At this point I won't be able to get back to version 3.0 of this addon until September. But Ember now has a [plan for the Octane release](https://blog.emberjs.com/2019/08/15/octane-release-plan.html)! (Roughly October 2019)

**Comment by Panman8201** [comment]
Again sorry for the delay, I _will_ be able to get to this starting next week and already have some rough code in a local (offline) fork.

**Comment by tavosansal** [comment]
Hey @Panman8201  ... I am wondering if there is a plan to release a new version with the update to Froala 3.0 without all of the other work that needs to be done on the ember side?
https://www.froala.com/wysiwyg-editor/docs/migrate-from-v2

The migration seems like not much has changed...and I am thinking all of the Ember stuff you wanna do (that sounds awesome btw) could take a while. We have been waiting to upgrade to 3.0 since it came out but we are still waiting on this addon to get updated. 

So yeah that was my question...just have a release in the meantime the ember upgrades get worked out.

Thanks.

**Comment by Panman8201** [comment]
@tavosansal Technically, yeah, it wouldn't be too hard to do (mainly just changing the initialization of it). I'm actually nearly done with the update (except docs). Just need the weekend to wrap things up and ensure I have the Octane patterns lined up. But it does also come at a cost of backwards compatibility (3.13 due to tracked properties). Are you needing to use it with an older version? I thought about also making a version that uses Ember Component for longer backwards compat.

**Comment by tavosansal** [comment]
@Panman8201 Oh wow...sounds like you got pretty far already! We are currently on Ember 3.12 so yeah, no tracked properties. It might be nice to have a longer compat version. I know we are probably not the only ones that will take us a bit to adopt the new style. Our codebase is pretty large.
I could help with that a bit if you give me some pointers. I just cloned the repo.

**Comment by Panman8201** [comment]
@tavosansal I've just published `3.0.0-alpha.1` with the v2 addon format (using Ember Component) but with Froala Editor v3. You can checkout the [`v2-with-froala-editor-3` branch](https://github.com/froala/ember-froala-editor/tree/v2-with-froala-editor-3) to see the differences. You'll probably want to lock at that specific version since I'll be releasing beta's with Glimmer Component and it'll probably have breaking changes.

_That felt good, probably should have done this long ago..._

**Comment by tavosansal** [comment]
Ryan (@Panman8201) Thank you so much for doing this. We appreciate it a ton.

**Comment by iurikothe** [comment]
+1 for Froala 3.x for Ember ;-) thanks guys!

**Comment by Emerson** [comment]
Thanks for putting this together, going to try this in our app soon. We found a bug in the latest version of 2.X that is preventing us from shipping a core feature and after contacting support, upgrading to version 3 is their suggested resolution.

**Comment by jrjohnson** [comment]
Tried out the alpha today worked super well. Drop in replacement with no changes needed (other than option differences between froala v2 and v3). 

Only thing is I am getting build warnings for both of our languages 

```
Warning: ignoring input sourcemap for node_modules/froala-editor/js/languages/fr.js because ENOENT: no such file or directory, open '/var/folders/g4/qxspl7qx4q77mxb5h1mh7n7w0000gq/T/broccoli-85586iJ9Zov7oMaDN/out-824-broccoli_debug_debug_3_vendor_js/node_modules/froala-editor/js/languages/fr.js.map'
Warning: ignoring input sourcemap for node_modules/froala-editor/js/languages/es.js because ENOENT: no such file or directory, open '/var/folders/g4/qxspl7qx4q77mxb5h1mh7n7w0000gq/T/broccoli-85586iJ9Zov7oMaDN/out-824-broccoli_debug_debug_3_vendor_js/node_modules/froala-editor/js/languages/es.js.map'
```

We this config:
```
'ember-froala-editor': {
      languages: ['fr','es'],
      themes: 'gray'
    },
```

The language files seem to be included (or at least they are working).

**Comment by Panman8201** [comment]
@jrjohnson humm... interesting. It's trying to pull in the map file for some reason, when there aren't any. They're working because the language files actually exist. What version of ember & ember-cli are  you on? Also, is there a larger stacktrace you can post?

**Comment by jrjohnson** [comment]
I'm not sure how to get more stack trace out of `ember serve`. That's the entirety of the output. 

It's reproducible with ember-cli `3.13.1` as:
```
ember new test-app
cd test app
npm install ember-froala-editor@3.0.0-alpha.1 --save
```

add 
```
'ember-froala-editor': {
      languages: ['fr','es'],
      plugins: false,
    },
```
to `ember-cli-build.js`

`npm start`

**Comment by jrjohnson** [comment]
I'm having some trouble with testing as well. 

We use `ember-cli-page-object` so I've always had a custom version of the `fillInFroalaEditor`. 

It doesn't seem to be possible any more to attach to the running editor to fill in / extract the value. 

Using `new FroalaEditor` on the same DOM node as the original just creates a second copy of the editor. I have a support conversation going with Froala about this and you can see what that that looks like in [this example](https://jsfiddle.net/jrjohnson/3v5buj2w/12/) Not sure if you're working around this somehow in the test-helper here that I just don't understand.

**Comment by Panman8201** [comment]
@jrjohnson Interesting, [I changed the test helper to basically do the same thing](https://github.com/froala/ember-froala-editor/blob/v2-with-froala-editor-3/addon-test-support/helpers/fill-in-froala-editor.js) and thought it was working properly but maybe not. If there is no way to get the editor instance otherwise, then yeah, that'll be a problem for any sort of testing. Realistically, it would be nice to mimic keyboard input into the editor, but I ran into problems when trying to do that previously. I don't see an open [issue on the Froala Editor repo](https://github.com/froala/wysiwyg-editor/issues), so let me know what you hear back from your support ticket. I'll also try to reproduce the sourcemap error, but I suspect it's an upstream issue since nothing  has really changed with the importing process of this addon...

In other news, Glimmer Component work is coming along in the [v3 branch](https://github.com/froala/ember-froala-editor/tree/v3). Just a few things left to add and then I should be able to cut the first beta. (I'll keep Ember Component stuff as alphas.)

**Comment by jrjohnson** [comment]
Was able to track down the source map warning cause and added a comment to https://github.com/froala/wysiwyg-editor/issues/3665. TLDR there is a `//# sourceMappingURL=es.js.map` reference at the end of each language file, but that map does not exist because the files are not actually minimized.

**Comment by Panman8201** [comment]
Awesome, thanks for the detective work. Let me know what you hear back about the double editor with initializing on the same element (using the test helper). That's the last bit for me to work through with the v3 branch, then I'll be releasing betas. (This week sometime.)

**Comment by jrjohnson** [comment]
Finally worked out a way to get into the running editor instance with support. It's a bit ugly, but you can see what they suggest here: https://jsfiddle.net/ilyasfroaal/p7ahjqb9/1/ 

In my page object helper I have access to the `DOMElement` and ember adds an ID to every classic component so my resulting helper code looks like:

```
getEditorInstance(element){
  return new Promise(resolve => {
    later(() => {
      const { INSTANCES } = FroalaEditor;
      const ourInstance = INSTANCES.find(instance => {
        const instanceElement = instance['$oel'][0];
        return instanceElement.id = element.id;
      });
      resolve(ourInstance);
    });
  });
}
```

This seems to be the only way to call `html.set()` and have it trigger the `contentChanged` event (and anything else that is bound on the original editor). I'm not sure if this will play well with CSS selectors and it's probably going to be an issue when moving to glimmer components which do not have an ID attribute automatically.

**Comment by dbollinger** [comment]
> `npm install ember-froala-editor@3.0.0-alpha.1 --save-dev`

Is this still the latest alpha, and is there anywhere I can go to follow along on this process? I'm planning to try this in my app soon, because it's my last dependency on jQuery 🥳

Anything I can do to help other than provide feedback when I try the alpha?

**Comment by Panman8201** [comment]
@dbollinger yes, that alpha is the only release with v3 ATM. I'm focusing releasing a beta today or tomorrow, working through the test helper issue that @jrjohnson highlighted above, then checking on maximizing backwards compatibility. What version of ember is everyone using here?

**Comment by iurikothe** [comment]
> 1. **What version of ember is everyone using here?**

Ember: 3.7.1

**Comment by dbollinger** [comment]
> What version of ember is everyone using here?

I have apps ranging from 3.4 to 3.14 -- for me personally, support for anything older than 3.8 would be "nice-to-have".

**Comment by dbollinger** [comment]
> Also, what is everyone's thoughts on requiring @content to be a SafeString from htmlSafe()

My instinct is to agree with this, but maybe a dev assertion would be a good starting point? I'm pretty sure my team's implementation wraps the component and enforces safe stings.

**Comment by Panman8201** [comment]
Alright, version `3.0.0-beta.1` is out! Unfortunately I'm waiting on a `setComponentTemplate()` polyfill to support anything prior to 3.13. 😢 But I *think* this should work back to `2.18` once that is out. Take a look at the [README for details on what's changed](https://github.com/froala/ember-froala-editor#upgrading-from-2x). The docs site is **not** updated at this point, next task is to tackle `ember-cli-addon-docs`.

@jrjohnson I was able to make the test helpers not even need the editor instance by creating a wrapper around the build-in `fillIn()` helper. It was actually easier/simpler than I had expected. But I *did* also work through the [access to the editor](https://github.com/froala/ember-froala-editor/commit/b3a6be273d38cace10f4049bc744fadf5db3caf1) issue with the example you provided.

**Comment by Panman8201** [comment]
Forgot to mention, very open to feedback, none of this is set in stone.

**Comment by jrjohnson** [comment]
That's fantastic @Panman8201! Love it.

**Comment by Panman82** [comment]
Latest update, I'm waiting to get the backwards compatibility stuff in before releasing 3.0.0 proper. Still waiting on that `setComponentTemplate()` polyfill, which will drive how far back compatibility goes. At this point I have it working back to 3.4, which is where the [`ember-native-class-polyfill`](https://github.com/pzuraq/ember-native-class-polyfill) support ends.

FWIW, the back-compat PR #82 is green but I don't want folks to have to [re-export the template when extending `<FroalaEditor>`](https://github.com/froala/ember-froala-editor/blob/v3-backwards-compat/tests/dummy/app/templates/components/extended-froala-editor.js#L1). So that's where `setComponentTemplate()` comes in...

**Comment by Panman82** [comment]
FYI, just finished a couple releases;

* `v3.0.0-alpha.2` to fix the `fillInFroalaEditor()` test helper
* `v3.0.0-beta.4` to include backwards compatibility back to ember 3.4

Folks who need to stick with v2 of this addon but want to use Froala Editor v3 can use the alpha. Otherwise work your way towards the betas, which use Ember Octane methodologies and specifically Glimmer Components.

**Notes with beta 4;**
1. When extending the `<FroalaEditor />` you'll need to re-export the template from your app (see the [example in the README](https://github.com/froala/ember-froala-editor/tree/v3.0.0-beta.4#extending-froalaeditor-component)) until the `setComponentTemplate()` polyfill is available
2. If you're upgrading from v2 and using ember < 3.10 you'll want to run the default blueprint (if not doing `ember install ember-froala-editor`) to get the other required polyfills. `ember g ember-froala-editor`

I'm going to close this case since v3 compatibility is now in releases. Feel free to open a new case for specific issues and/or discussion, I'd be happy to chat.

**Comment by Srinivasan-Natarajan** [comment]
i am using ember-froala-editor v3.0.0-alpha.2 with froala-editor 3.1.1 ..............application runs after removing jquery as mentioned in the tag release notes like below

$.FroalaEditor.RegisterCommand('myButton', {}); 
to 
FroalaEditor.RegisterCommand('myButton', {}); 

but when i run the test cases which throws error like  
bootstrap.js:127 Uncaught TypeError: Cannot read property 'DEFAULTS' of undefined

got error on below line in ember-froala-editor
if (Ember.$.FroalaEditor.DEFAULTS.hasOwnProperty(propertyName)) {

**Comment by Srinivasan-Natarajan** [comment]
i am using froala 3.1.1 and ember-froala v3.0.0.alpha-2.
error occurred while getting instance on didUpdateAttrs hook. only partial instance is seems to be loaded... editor.html return undefined

**Comment by Panman82** [comment]
@Srinivasan-Natarajan The 3.x alphas here actually have Froala Editor 2.x, they were a way for folks to use Froala Editor 2 with Glimmer Components in Ember. I'd suggest using the latest beta here for now (which has Froala Editor 3) until we can release a stable version (the latest beta is "stable" at this point).

**Comment by Srinivasan-Natarajan** [comment]
@Panman82 hi we are still using older way to extending the component. will it work on beta?

import FroalaEditorComponent from 'ember-froala-editor/components/froala-editor';
export default FroalaEditorComponent.extend({
  options: {},
  'on-blur': function() {}
});

**Comment by Panman82** [comment]
@Srinivasan-Natarajan Yes, you are still able to apply default options and event handlers, but the `export` line has changed since it comes from the Glimmer Component class now. See "Extending `<FroalaEditor>` component" in the README: https://github.com/froala/ember-froala-editor#extending-froalaeditor-component
